pkg/spacebinding/spacebindinglister_test.go

```diff
@@ -0,0 +1,238 @@
+		// we have a parent space
+		spaceA := spacetest.NewSpace(test.HostOperatorNs, "Space-A",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithLabel(toolchainv1alpha1.SpaceCreatorLabelKey, "catwomen"),
+		)
+		spaceB := spacetest.NewSpace(test.HostOperatorNs, "Space-B",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(spaceA.GetName()),
+		)
+		spaceD := spacetest.NewSpace(test.HostOperatorNs, "Space-D",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(spaceB.GetName()),
+		)
+		spaceE := spacetest.NewSpace(test.HostOperatorNs, "Space-E",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(spaceD.GetName()),
+		)
+
+		spaceC := spacetest.NewSpace(test.HostOperatorNs, "Space-C",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(spaceA.GetName()),
+		)
+		spaceF := spacetest.NewSpace(test.HostOperatorNs, "Space-F",
+			spacetest.WithTierName("advanced"),
+			spacetest.WithSpecTargetCluster(test.MemberClusterName),
+			spacetest.WithSpecParentSpace(spaceC.GetName()),
+		)
+		errorspace := spacetest.NewSpace(test.HostOperatorNs, "errorspace",
+			spacetest.WithSpecParentSpace("invalidparent"))
+		errorspacebindings := spacetest.NewSpace(test.HostOperatorNs, "errorspacebindings")
+		spaces := map[string]*toolchainv1alpha1.Space{
+			spaceA.GetName(): spaceA,
+			spaceB.GetName(): spaceB,
+			spaceC.GetName(): spaceC,
+			spaceD.GetName(): spaceD,
+			spaceE.GetName(): spaceE,
+			spaceF.GetName(): spaceF,
+		}
+		// the getSpaceFunc returns the space based on the given name
+		getSpaceFunc := func(spaceName string) (*toolchainv1alpha1.Space, error) {
+			if space, found := spaces[spaceName]; found {
+				return space, nil
+			}
+			if spaceName == "invalidparent" {
+				return nil, errors.New("mock error")
+			}
+			return nil, fmt.Errorf("space not found: %s", spaceName)
+		}
+
+		// listParentSpaceBindingFunc returns the specific spacebindings for the given space
+		listParentSpaceBindingFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
+			switch spaceName {
+			case spaceA.GetName():
+				return []toolchainv1alpha1.SpaceBinding{
+					*spacebinding.NewSpaceBinding(catwomenMur, spaceA, spaceA.Name, spacebinding.WithRole("admin")),
+				}, nil
+			case spaceB.GetName():
+				return []toolchainv1alpha1.SpaceBinding{
+					*spacebinding.NewSpaceBinding(batmanMur, spaceB, spaceB.Name, spacebinding.WithRole("maintainer")),
+				}, nil
+			case spaceC.GetName():
+				return []toolchainv1alpha1.SpaceBinding{
+					*spacebinding.NewSpaceBinding(batmanMur, spaceC, spaceC.Name, spacebinding.WithRole("maintainer")),
+				}, nil
```

Comment:
what is the benefit (for the purpose of the test) of having two spaces with the same SpaceBindings (maintainer for batmat) and inheriting from the same parent space?
Wouldn't make more sense to either drop one of them or let spaceC inherit from spaceB and expect no change there even though there is another SpaceBinding?

Reply:
Good point. Let's have spaceC inheriting from spaceB and expect no change, this is not covered as usecase, even if it doesn't make much sense to create a spacebinding with a role that the user already has (`maintainer` in this case) , but someone might create one by mistake and things should still work.